File: Clocktower.Server/Data/GameState.cs
namespace Clocktower.Server.Data;

public class GameState : IIdentifiable
{
    public string Id { get; set; }
    public string GuildId { get; set; }
    public List<GameUser> Users { get; set; } = [];
    public IEnumerable<GameUser> Players => Users.Where(o => o.UserType == UserType.Player);
    public IEnumerable<GameUser> StoryTellers => Users.Where(o => o.UserType == UserType.StoryTeller);
    public IEnumerable<GameUser> Spectators => Users.Where(o => o.UserType == UserType.Spectator);
    public int MaxPlayers { get; set; } = 0;

    public GameUser CreatedBy { get; set; }
    public DateTime CreatedDate { get; set; }

    public GameTime GameTime { get; set; }

    public bool IsFull => MaxPlayers != 0 && Players.Count() >= MaxPlayers;
}

File: Clocktower-UI/src/types/gameState.ts
import type {ClocktowerServerDataGameState} from "@/api";
import {type GameUser, mapToGameUser} from "@/types/gameUser.ts";

export type GameState = {
    id: string;
    guildId: string,
    players: GameUser[];
    spectators: GameUser[];
    storyTellers: GameUser[];
    createdDate: Date,
    createdBy: GameUser
}

export function mapToGameState(apiGame: ClocktowerServerDataGameState): GameState {
    return {
        id: apiGame.id!,
        guildId: apiGame.guildId!,
        players: apiGame.players!.map(mapToGameUser),
        spectators: apiGame.spectators!.map(mapToGameUser),
        storyTellers: apiGame.storyTellers!.map(mapToGameUser),
        createdDate: apiGame.createdDate!,
        createdBy: mapToGameUser(apiGame.createdBy!)
    };
}